src/routes/login.ts

```diff
@@ -0,0 +1,54 @@
+import { Type } from '@sinclair/typebox'
+import { UserInputError } from 'apollo-server-express'
+import bcrypt from 'bcrypt'
+import { Request, Response } from 'express'
+import { authCookie } from '../authenticateRequest'
+import { trimAll } from '../input-validation/trimAll'
+import { validateWithJSONSchema } from '../input-validation/validateWithJSONSchema'
+import UserAccount from '../models/user_account'
+import { passwordInput, usernameInput } from './register'
+
+const loginInput = Type.Object(
+  {
+    username: usernameInput,
+    password: passwordInput,
+  },
+  { additionalProperties: false },
+)
+
+const validateLoginInput = validateWithJSONSchema(loginInput)
+
+const login =
+  (penaltySeconds = 10) =>
+  async (request: Request, response: Response) => {
+    const valid = validateLoginInput(trimAll(request.body))
+    if ('errors' in valid) {
+      return response
+        .status(400)
+        .json(new UserInputError('Login input invalid', valid.errors))
+        .end()
+    }
+
+    const user = await UserAccount.findOne({
+      where: {
+        username: valid.value.username,
+      },
+    })
+    if (user === null) {
+      // Penalize
+      await new Promise((resolve) => setTimeout(resolve, penaltySeconds * 1000))
+      return response.status(401).end()
+    }
+    if (!bcrypt.compareSync(valid.value.password, user.passwordHash)) {
+      // Penalize
+      await new Promise((resolve) => setTimeout(resolve, penaltySeconds * 1000))
+      return response.status(401).end()
+    }
```

Comment:
**though (non-blocking):** this means that if I type the wrong username or password, I'll have to wait 10 seconds until I see an error message, right? I think that could become annoying from a UX perspective. I wonder if we could instead throttle access to that endpoint based on the user's IP (not necessarily in this PR though)

Reply:
Yes, a more advanced feature of this would be remembering the user, however, the point here is to slow down attackers, which most likely will switch IP addresses. I think a better solution will be to implement a CAPTCHA, like https://friendlycaptcha.com/.

Reply:
I like the CAPTCHA solution 👍